src/controller/robot/controller.go

```diff
@@ -0,0 +1,343 @@
+	if err := d.setProjectID(ctx, r); err != nil {
+		return 0, err
+	}
+
+	if r.ExpiresAt == 0 {
+		tokenDuration := time.Duration(config.RobotTokenDuration()) * time.Minute
+		r.ExpiresAt = time.Now().UTC().Add(tokenDuration).Unix()
+	}
+
+	key, err := config.SecretKey()
+	if err != nil {
+		return 0, err
+	}
+	str := utils.GenerateRandomString()
+	secret, err := utils.ReversibleEncrypt(str, key)
+	if err != nil {
+		return 0, err
+	}
+
+	robotID, err := d.robotMgr.Create(ctx, &model.Robot{
+		Name:        r.Name,
+		Description: r.Description,
+		ProjectID:   r.ProjectID,
+		ExpiresAt:   r.ExpiresAt,
+		Secret:      secret,
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	if err := d.createPermission(ctx, r); err != nil {
+		return 0, err
+	}
+	return robotID, nil
+}
+
+// Delete ...
+func (d *controller) Delete(ctx context.Context, id int64) error {
+	if err := d.robotMgr.Delete(ctx, id); err != nil {
+		return err
+	}
+	if err := d.rbacMgr.DeletePermissionsByRole(ctx, ROBOTTYPE, id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Update ...
+func (d *controller) Update(ctx context.Context, r *Robot) error {
+	if r == nil {
+		return errors.New("cannot update a nil robot").WithCode(errors.BadRequestCode)
+	}
+	if err := d.robotMgr.Update(ctx, &r.Robot); err != nil {
+		return err
+	}
+	if err := d.setProjectID(ctx, r); err != nil {
+		return err
+	}
+	// update the permission
+	if err := d.rbacMgr.DeletePermissionsByRole(ctx, ROBOTTYPE, r.ID); err != nil {
+		return err
+	}
+	if err := d.createPermission(ctx, r); err != nil {
+		return err
+	}
+	return nil
+}
+
+// List ...
+func (d *controller) List(ctx context.Context, query *q.Query, option *Option) ([]*Robot, error) {
+	robots, err := d.robotMgr.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	var robotAccounts []*Robot
+	for _, r := range robots {
+		rb, err := d.populate(ctx, r, option)
+		if err != nil {
+			return nil, err
+		}
+		robotAccounts = append(robotAccounts, rb)
+	}
+	return robotAccounts, nil
+}
+
+func (d *controller) createPermission(ctx context.Context, r *Robot) error {
+	if r == nil {
+		return nil
+	}
+
+	for _, per := range r.Permissions {
+		policy := &rbac_model.PermissionPolicy{}
+		scope, err := d.toScope(ctx, per)
+		if err != nil {
+			return err
+		}
+		policy.Scope = scope
+
+		for _, access := range per.Access {
+			policy.Resource = access.Resource.String()
+			policy.Action = access.Action.String()
+			policy.Effect = access.Effect.String()
+
+			policyID, err := d.rbacMgr.CreateRbacPolicy(ctx, policy)
+			if err != nil {
+				return err
+			}
+
+			_, err = d.rbacMgr.CreatePermission(ctx, &rbac_model.RolePermission{
+				RoleType:           ROBOTTYPE,
+				RoleID:             r.ID,
+				PermissionPolicyID: policyID,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (d *controller) populate(ctx context.Context, r *model.Robot, option *Option) (*Robot, error) {
```

Comment:
Fetch or attach is better than populate, on auth mod, the populate often involves data migrate, such as populateGroup, populateUser, current case only retrieves the information.